R/interpolate.R

```diff
@@ -118,3 +117,86 @@ sqlCommentSpec <- function(start, end, endRequired) {
 sqlQuoteSpec <- function(start, end, escape = "", doubleEscape = TRUE) {
   list(start, end, escape, doubleEscape)
 }
+
+#' @export
+#' @rdname sqlParseVariables
+sqlParseVariablesImpl <- function(sql, quotes, comments) {
+  sql_arr <- strsplit(as.character(sql), "", fixed=T)[[1]]
+  var_pos_start <- integer()
+  var_pos_end <- integer()
+
+  # prepare comments
+  for(c in seq_along(comments)) {
+    comments[[c]][[1]] <- strsplit(comments[[c]][[1]], "", fixed=T)[[1]]
+    comments[[c]][[2]] <- strsplit(comments[[c]][[2]], "", fixed=T)[[1]]
+  }
+  
+  var_chars <- c(LETTERS, tolower(LETTERS), 0:9, "_")
+  in_quote <- 0L
+  in_comment <- 0L
+  i <- 1
+  while(i < length(sql_arr)) {
+    # only check for variables if neither commented nor quoted
+    if (in_quote == 0L && in_comment == 0L) {
+      if (sql_arr[[i]] == "?") {
+        # consume everything alphanumeric and _ up to end of variable
+        var_pos_start <- c(var_pos_start, i)
+        repeat {
+          i <- i + 1
+          if (i > length(sql_arr) || !(sql_arr[[i]] %in% var_chars)) {
+            break
+          }
+        }
+        var_pos_end <- c(var_pos_end, i-1)
+        next
+      }
+    }
+    # only check for quoted strings when not in commented section
+    if (in_comment == 0L) {
+      # check all quote defintions, they can only be single characters
+      for(q in seq_along(quotes)) {
+        if (in_quote == 0L) {
+          quote_start_char <- quotes[[q]][[1]]
+          if (identical(sql_arr[[i]], quote_start_char))  {
+            in_quote <- q
+            break
+          }
+        } else {
+          # only check the end of the active quote definition
+          # TODO: support end quote escaping (e.g. \")
```

Comment:
I think there was support for it, although not tested: https://codecov.io/github/rstats-db/DBI/src/sqldelim.cpp?ref=ef1a72121ed1f34e5bf94a913ea3410d9b289f42#l-61 . I'd like to review the old code and have it fully covered by tests before proceeding here. Also, to me it looks like the construct `"SELECT ? FROM A"` will be interpolated in a strange way by the C++ implementation.